src/subclass/widget.rs

```diff
@@ -0,0 +1,249 @@
+            T: WidgetImpl,
+        {
+            let instance = &*(ptr as *mut T::Instance);
+            let imp = instance.get_impl();
+            let wrap: Widget = from_glib_borrow(ptr);
+
+            imp.adjust_baseline_allocation(&wrap, &mut *baseptr)
+        }
+
+        unsafe extern "C" fn widget_adjust_baseline_request<T: ObjectSubclass>(
+            ptr: *mut gtk_sys::GtkWidget,
+            minptr: *mut i32,
+            natptr: *mut i32,
+        ) where
+            T: WidgetImpl,
+        {
+            let instance = &*(ptr as *mut T::Instance);
+            let imp = instance.get_impl();
+            let wrap: Widget = from_glib_borrow(ptr);
+
+            imp.adjust_baseline_request(&wrap, &mut *minptr, &mut *natptr)
+        }
+
+        unsafe extern "C" fn widget_adjust_size_allocation<T: ObjectSubclass>(
+            ptr: *mut gtk_sys::GtkWidget,
+            orientation: gtk_sys::GtkOrientation,
+            minptr: *mut i32,
+            natptr: *mut i32,
+            posptr: *mut i32,
+            sizeptr: *mut i32,
+        ) where
+            T: WidgetImpl,
+        {
+            let instance = &*(ptr as *mut T::Instance);
+            let imp = instance.get_impl();
+            let wrap: Widget = from_glib_borrow(ptr);
+
+            let wrap_orientation: Orientation = match orientation {
+                0 => Orientation::Horizontal,
+                1 => Orientation::Vertical,
+                num => Orientation::__Unknown(num),
+            };
+
+            imp.adjust_size_allocation(&wrap, wrap_orientation, &mut *minptr, &mut *natptr, &mut *posptr, &mut *sizeptr)
+        }
+
+        unsafe extern "C" fn widget_adjust_size_request<T: ObjectSubclass>(
+            ptr: *mut gtk_sys::GtkWidget,
+            orientation: gtk_sys::GtkOrientation,
+            minptr: *mut i32,
+            natptr: *mut i32,
+        ) where
+            T: WidgetImpl,
+        {
+            let instance = &*(ptr as *mut T::Instance);
+            let imp = instance.get_impl();
+            let wrap: Widget = from_glib_borrow(ptr);
+
+            let wrap_orientation: Orientation = match orientation {
+                0 => Orientation::Horizontal,
+                1 => Orientation::Vertical,
+                num => Orientation::__Unknown(num),
+            };
+
+            imp.adjust_size_request(&wrap, wrap_orientation, &mut *minptr, &mut *natptr)
+        }
+
+        unsafe extern "C" fn widget_button_press_event<T: ObjectSubclass>(
+            ptr: *mut gtk_sys::GtkWidget,
+            btnptr: *mut gdk_sys::GdkEventButton,
+        ) -> glib_sys::gboolean
+            where T: WidgetImpl,
+        {
+            let instance = &*(ptr as *mut T::Instance);
+            let imp = instance.get_impl();
+            let wrap: Widget = from_glib_borrow(ptr);
+            let mut evwrap: gdk::EventButton = from_glib_borrow(btnptr);
+
+            imp.button_press_event(&wrap, &mut evwrap) as glib_sys::gboolean
```

Comment:
This is unsafe: you can't pass a mutable reference to something retrieved via `from_glib_borrow()` as that allows to produce double frees

Reply:
I.e. you must only pass immutable references if you use `from_glib_borrow()`

Reply:
Is there a way that could allow `EventButton` to be mutable? Or will it need to be immutable?

Reply:
Hmm, I don't see any clean way right now

Reply:
The event stuff will all be nicer with GTK 4 anyway :)